packages/router/contracts/adapters/PendleAdapter.sol

```diff
@@ -0,0 +1,387 @@
+      IMarginlyRouter(data.initiator).adapterCallback(msg.sender, uint256(-ptToAccount), data.data);
+    } else {
+      revert();
+    }
+  }
+
+  /// @dev callbacks are used in exactOutput swaps only
+  function uniswapV3SwapCallback(int256 amount0Delta, int256 amount1Delta, bytes calldata _data) external {
+    CallbackData memory data = abi.decode(_data, (CallbackData));
+    (address tokenIn, address tokenOut) = (data.tokenIn, data.tokenOut);
+    PoolData memory poolData = getPoolDataSafe(tokenIn, tokenOut);
+    require(msg.sender == poolData.uniswapV3LikePool);
+
+    // TODO redefine conditions and requires inside
+    if (amount0Delta > 0) {
+      // this clause is realized in case of exactOutput with pt tokens as output
+      // we need to send tokenIn to uniswapV3 to finalize both tokenIn -> sy and sy -> pt swaps
+      require(tokenIn < data.syToken);
+      IMarginlyRouter(data.initiator).adapterCallback(msg.sender, uint256(amount0Delta), data.data);
+    } else if (amount1Delta > 0) {
+      // this clause is realized in case of exactOutput with pt tokens as input
+      // we need to trigger pendle pt -> sy exactOutput swap
+      require(tokenIn < data.syToken);
+      _pendleApproxSwapPtForExactSy(
+        getMarketData(poolData),
+        poolData.uniswapV3LikePool,
+        uint256(amount1Delta),
+        data.maxAmountIn,
+        _data
+      );
+    } else {
+      revert();
+    }
+
+    (bool isExactInput, uint256 amountToPay) = amount0Delta > 0
+      ? (tokenIn < data.syToken, uint256(amount0Delta))
+      : (data.syToken < tokenIn, uint256(amount1Delta));
+
+    require(isExactInput);
+
+    IMarginlyRouter(data.initiator).adapterCallback(msg.sender, amountToPay, data.data);
+  }
+
+  function getPoolDataSafe(address tokenA, address tokenB) private view returns (PoolData memory poolData) {
+    poolData = getPoolData[tokenA][tokenB];
+    if (poolData.pendleMarket == address(0)) revert();
+  }
+
+  function getMarketData(PoolData memory poolData) private view returns (PendleMarketData memory) {
+    IPMarket market = IPMarket(poolData.pendleMarket);
+    (IStandardizedYield sy, IPPrincipalToken pt, IPYieldToken yt) = market.readTokens();
+    return PendleMarketData({market: market, sy: sy, pt: pt, yt: yt, uniswapV3: poolData.uniswapV3LikePool});
+  }
+
+  function _swapExactInputPreMaturity(
+    PendleMarketData memory marketData,
+    address recipient,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountIn,
+    uint256 minAmountOut,
+    bytes calldata data
+  ) private returns (uint256 amountOut) {
+    uint256 syAmountOut;
+    if (tokenIn == address(marketData.pt)) {
+      // pt to pendle -> sy to uniswap -> tokenOut to recipient
+      IMarginlyRouter(msg.sender).adapterCallback(address(marketData.market), amountIn, data);
+      (syAmountOut, ) = marketData.market.swapExactPtForSy(marketData.uniswapV3, amountIn, '');
+      (, amountOut) = _uniswapV3LikeSwap(
```

Comment:
I use the following names of token types:

- SY - Standardized Yield (name from Pendle)
- PT - Principal Token (name from Pendle)
- YQT - Yield Quote Token (suggested name)
- QT - Quote Token (suggested name)

Pools contains tokens:
1) Pendle pool: PT/SY
2) External additional pool: YQT/QT

Exchange rate: 1.0 SY = 1.0 YQT

At this line you have only SY token, but secondary pool requires YQT.
Conversation from SY to YQT is required.